In [ ]:
url_api="https://ressources.data.sncf.com/api/v2/catalog/datasets/regularite-mensuelle-ter/exports/json?limit=-1&offset=0&timezone=UTC"


In [ ]:
import requests
import pandas as pd
#recupération des données de régularité via l'API SNCF 
req = requests.get(url_api)
wb = req.json()

# normalisation du fichier json
df = pd.json_normalize(wb)
print(df.shape)
df.head(8) #inspection des 8 premières lignes


(1825, 9)


date              region  nombre_de_trains_programmes  \
0  2013-01     Basse Normandie                       3331.0   
1  2013-01   Champagne Ardenne                       6648.0   
2  2013-01            Lorraine                          NaN   
3  2013-02           Aquitaine                       7926.0   
4  2013-02  Nord Pas de Calais                      17478.0   
5  2013-02         Rhône Alpes                      28215.0   
6  2013-03              Alsace                      19138.0   
7  2013-03           Aquitaine                       8167.0   

   nombre_de_trains_ayant_circule  nombre_de_trains_annules  \
0                          3297.0                      34.0   
1                          6595.0                      53.0   
2                             NaN                       NaN   
3                          7832.0                      94.0   
4                         17228.0                     250.0   
5                         27910.0                     305.0   
6                         18970.0                     168.0   
7                          8081.0                      86.0   

   nombre_de_trains_en_retard_a_l_arrivee  taux_de_regularite  \
0                                   311.0           90.567182   
1                                   334.0           94.935557   
2                                     NaN                 NaN   
3                                   912.0           88.355465   
4                                  1040.0           93.963316   
5                                  3771.0           86.488714   
6                                   658.0           96.531365   
7                                   822.0           89.827992   

   nombre_de_trains_a_l_heure_pour_un_train_en_retard_a_l_arrivee  \
0                                           9.601286                
1                                          18.745509                
2                                                NaN                
3                                           7.587719                
4                                          15.565385                
5                                           6.401220                
6                                          27.829787                
7                                           8.830900                

                                        commentaires  
0  Grand froid et épisode neigeux les semaines 3 ...  
1  Conditions météorologiques dégradées du 15 au ...  
2  Le Président de la Région Lorraine s'est oppos...  
3     Inondations au sud de Dax du 11 au 13 février.  
4                            Intempéries mi-février.  
5                     Météo difficile et intrusions.  
6                                               None  
7  Coup de froid sur l'ensemble de la région lors...

#**Description de la base**

Ce jeu de données porte sur la régularité des trains de la SNCF depuis janvier 2015 jusqu'au 28 octobre 2022. Il contient 1825 observations qui sont enrégistrées mensuellement et par région. Il contient 9 variables à savoir:

date	
region
nombre_de_trains_programmes
nombre_de_trains_ayant_circule
nombre_de_trains_annules
nombre_de_trains_en_retard_a_l_arrivee
taux_de_regularite	nombre_de_trains_a_l_heure_pour_un_train_en_retard_a_l_arrivee
commentaires


#**Analyse de la qualité de base en vue du traitement des données**


In [ ]:
#Type des variables
df.dtypes # print pour observer 

date                                                               object
region                                                             object
nombre_de_trains_programmes                                       float64
nombre_de_trains_ayant_circule                                    float64
nombre_de_trains_annules                                          float64
nombre_de_trains_en_retard_a_l_arrivee                            float64
taux_de_regularite                                                float64
nombre_de_trains_a_l_heure_pour_un_train_en_retard_a_l_arrivee    float64
commentaires                                                       object
dtype: object

In [ ]:
#quelques statistiques sur les variables quantitatives
df.describe()

nombre_de_trains_programmes  nombre_de_trains_ayant_circule  \
count                  1788.000000                     1788.000000   
mean                  12292.477629                    12039.897651   
std                    9015.810259                     8818.687906   
min                    1129.000000                     1125.000000   
25%                    5901.500000                     5825.750000   
50%                    9474.000000                     9263.500000   
75%                   14888.000000                    14406.500000   
max                   46329.000000                    45569.000000   

       nombre_de_trains_annules  nombre_de_trains_en_retard_a_l_arrivee  \
count               1788.000000                             1788.000000   
mean                 258.276846                             1011.242729   
std                  291.916890                              837.461109   
min                    4.000000                               35.000000   
25%                   83.000000                              421.000000   
50%                  157.000000                              729.500000   
75%                  333.000000                             1239.500000   
max                 4024.000000                             4576.000000   

       taux_de_regularite  \
count         1788.000000   
mean            91.574464   
std              3.420634   
min             74.365711   
25%             89.670614   
50%             92.017786   
75%             94.124250   
max             98.026494   

       nombre_de_trains_a_l_heure_pour_un_train_en_retard_a_l_arrivee  
count                                        1788.000000               
mean                                           13.044802               
std                                             6.218286               
min                                             2.901025               
25%                                             8.681130               
50%                                            11.527859               
75%                                            16.019116               
max                                            49.671233

In [ ]:
#Donnnées manquantes
df.isnull().sum()

date                                                                0
region                                                              0
nombre_de_trains_programmes                                        37
nombre_de_trains_ayant_circule                                     37
nombre_de_trains_annules                                           37
nombre_de_trains_en_retard_a_l_arrivee                             37
taux_de_regularite                                                 37
nombre_de_trains_a_l_heure_pour_un_train_en_retard_a_l_arrivee     37
commentaires                                                      899
dtype: int64

In [ ]:
#Identification des régions présentes dans la base
list_regions=df['region'].unique()  #list des régions sous forme numpy.ndarray
print(len(list_regions))
print(list_regions)
#Nombre d'année couvert par la base
list_annee=df['date'].unique()  #list des années sous forme numpy.ndarray
print(len(list_annee))
print(list_annee)

28
['Basse Normandie' 'Champagne Ardenne' 'Lorraine' 'Aquitaine'
 'Nord Pas de Calais' 'Rhône Alpes' 'Alsace' 'Picardie' 'Franche Comté'
 'Midi Pyrénées' 'Bourgogne' 'Bretagne' 'Limousin' 'Pays-de-la-Loire'
 'Languedoc Roussillon' 'Auvergne' 'Haute Normandie' 'Poitou Charentes'
 'Centre' "Provence Alpes Côte d'Azur" 'Grand Est' 'Hauts-de-France'
 'Normandie' 'Occitanie' 'Bourgogne-Franche-Comté' 'Auvergne-Rhône-Alpes'
 'Centre Val-de-Loire' 'Nouvelle Aquitaine']
119
['2013-01' '2013-02' '2013-03' '2013-04' '2013-05' '2013-06' '2013-07'
 '2013-08' '2013-09' '2013-10' '2013-11' '2013-12' '2014-01' '2014-02'
 '2014-03' '2014-04' '2014-05' '2014-06' '2014-07' '2014-08' '2014-09'
 '2014-10' '2014-11' '2014-12' '2015-01' '2015-02' '2015-03' '2015-04'
 '2015-05' '2015-06' '2015-07' '2015-08' '2015-09' '2015-10' '2015-11'
 '2015-12' '2016-01' '2016-02' '2016-03' '2016-04' '2016-05' '2016-06'
 '2016-07' '2016-08' '2016-09' '2016-10' '2016-11' '2016-12' '2017-01'
 '2017-02' '2017-03' '2017-04' '

On remarque qu'il ya 28 régions présentes tandis que la france compte actuellement et officiellement 12 régions. Le problème se trouve au niveau de l'évolution du découpage administratif de 2013 à 2022

In [ ]:
df['date']=pd.to_datetime(df['date'])
print(type(df['date'][0]))
print(df['date'][0])
df['annee'] = df['date'].dt.year 
df['mois'] = df['date'].dt.month 
#df.loc[df['column name'] condition, 'new column name'] = 'value if condition is met'

<class 'pandas._libs.tslibs.timestamps.Timestamp'>
2013-01-01 00:00:00


In [ ]:
df_l=df[df['region']=='Lorraine']

